### DISCUSSION TITLE
Docs: clarify if `Result.all() -> Sequence[[Row]]` returns an iterator or a (potentially large) list

### DISCUSSION THREAD

**Discussion by jgehrcke** [sqlalchemy/sqlalchemy#9735] [Usage Questions] [answered]
For processing a potentially large number of rows returned by a query I was wondering how the following idiom behaves in terms of memory consumption:

```python
results = Session.scalars(query_statement).all()
for result in results:
   ...
```

I was hoping that `results` is an iterator, and I think it is, i.e. `results` is never going to be a big object in memory. Only `materialized = list(results)` may be large.

I was trying to get confirmation from https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Result.all but there I found the statement

> "Return all rows in a list"

which maybe needs clarification. Super happy to provide a patch, but yeah I thought I start this discussion first because I am after all not really familiar with SQLAlchemy.

**Comment by CaselIT** [comment, marked as answer]
Hi,

I don't think it needs clarification since an iterator is not a sequence since sequence implies that all elements are present in the container.
That statement is correct, since `all()` returns a list currently.

You may be looking for https://docs.sqlalchemy.org/en/20/core/connections.html#using-server-side-cursors-a-k-a-stream-results

**Comment by jgehrcke** [comment]
Ah! Glad I asked. I was searching the docs for "row iterator" but should have tried "streaming results"! The section you linked to is great, discussing buffer etc. Lovely. Maybe we want to add a sentence to the `all()` documentation, cross-linking the topics.

**Reply by CaselIT** [reply]
Sure, that may make sense
